### DISCUSSION TITLE
Understanding the tensor structure

### DISCUSSION THREAD

**Discussion by NicoClinco** [Cambridge-ICCS/FTorch#108] [Q&A] [answered]
Hello everybody,
i'm facing some difficulties in understanding what does it means 'dimension' when declaring a torch_tensor.
For example, if we consider the tutorial:

type(torch_tensor), dimension(1) :: in_tensor

Is it true that dimension is refered to the batch size or dimension means another thing?
For example, suppose that i have an MLP trained in python. I can pass a tensor of shape [batch_size,features] in python directly and the net will propagate. 
In fortran dimension is refered to the batchsize? Or i can also specify dimension=1 and then pass a 2d array in fortran?

Thanks in advance.
Nicola

**Comment by jatkinson1000** [comment, marked as answer]
Hi @NicoClinco I'm guessing you're coming from python and are less familiar with Fortran? The line you quote declares an array of tensors. This is because fortran is more restrictive than python as a compiled language, so we need to do this for models that may have multiple tensor inputs.

Breaking it down, the statement says: "create a vector of `torch_tensor` types containing one item, and name it (the vector) `in_tensor`." 

If you wanted to pass three different tensors to your model as inputs then you would declare it as `dimension(3)` etc.

Batch size would be part of the tensor you pass in as an element of this array. so if you had a single input to your net of length n, batch size m, you would declare this as above, and then set `in_tensor(1)` to be an nxm array. 

Does this make sense? Please ask if anything is unclear, and if you can point me at some code I'd be happy to help you get it working! 

Perhaps it would be clearer if we called this variable `in_tensors` in our tutorials?

**Comment by NicoClinco** [comment]
Apologize for the previous question, i'm newby with torch and indeed i have
never used a network with multiple inputs.
Thank you for the quick answer.
Nicola

Il giorno sab 6 apr 2024 alle ore 19:39 jatkinson1000 <
***@***.***> ha scritto:

**Reply by jatkinson1000** [reply]
No need to apologise - it's useful feedback to hear from users where we can improve our documentation/code.

Let us know if you have any other questions, and if you have any projects using FTorch we'd love to know and feature them.

If my reply above solved this for you I'd appreciate it if you could mark it as an answer, thanks!

**Reply by jatkinson1000** [reply]
I'm currently working on a new example that should make this a bit clearer using a net with multiple inputs. It is still WIP but you can take a look here https://github.com/Cambridge-ICCS/FTorch/pull/94

**Comment by jatkinson1000** [comment]
Hi @NicoClinco Further to this discussion I have opened a pull request #110 to try and clarify some of the aspects you raised here.
Please feel free to take a look and comment/review.
